### ISSUE TITLE
Custom Gradients

### ISSUE THREAD

**Issue by glenbhermon** [kunalbanerjee/softmax_alternatives#1]
Hello,
I was just going through the code, I found that while executing all your softmax alternatives except sm_softmax, they return a 'nan' value for the loss while training, however it runs when the custom gradient part of the code is removed (@tf.custom_gradient and def custom_grad(dy) parts). Also I am not able to understand why custom gradients are used?

**Comment by kunalbanerjee** [comment]
I am not sure why you are getting NaNs because I never encountered these and could train the models easily - perhaps it has something to do with Tensorflow version.
In some cases, for e.g., Taylor softmax, we want to use the same forward pass but while doing the backward pass, the calcultation for the gradient will differ based on whether the Taylor expansion is considered to be a finite series or an infinite one - hence custom gradients are used. Hope this clarifies your doubt.
